File: functions/src/services/notificationService.ts
import { db, messaging } from '../config/firebase';

export interface SendPushNotificationParams {
  userId: string;
  title: string;
  body: string;
  data?: Record<string, string>;
}

export async function sendPushNotificationToUser(params: SendPushNotificationParams): Promise<boolean> {
  try {
    console.log(`[sendPushNotificationToUser] Target userId: ${params.userId}`);
    const userDoc = await db.collection('users').doc(params.userId).get();
    if (!userDoc.exists) {
      console.warn(`[sendPushNotificationToUser] User doc ${params.userId} does not exist in Firestore.`);
      return false;
    }

    const userData = userDoc.data();
    const fcmToken = userData?.fcmToken || (Array.isArray(userData?.fcmTokens) ? userData.fcmTokens[0] : undefined);

    if (!fcmToken) {
      console.warn(`[sendPushNotificationToUser] No fcmToken field found in users/${params.userId}`);
      return false;
    }

    console.log(`[sendPushNotificationToUser] Sending FCM payload to token (${fcmToken.substring(0, 20)}...)...`);

    const response = await messaging.send({
      token: fcmToken,
      notification: {
        title: params.title,
        body: params.body,
      },
      android: {
        priority: 'high',
        notification: {
          channelId: 'high_importance_channel',
          priority: 'high',
          defaultSound: true,
        },
      },
      apns: {
        payload: {
          aps: {
            alert: {
              title: params.title,
              body: params.body,
            },
            sound: 'default',
          },
        },
      },
      data: params.data || {},
    });

    console.log(`[sendPushNotificationToUser] FCM Notification sent successfully! Response ID: ${response}`);
    return true;
  } catch (error: any) {
    console.error(`[sendPushNotificationToUser] Error sending FCM notification to user ${params.userId}:`, error?.message || error, error?.stack || '');
    return false;
  }
}

File: lib/core/services/notification/notification_service.dart
import 'dart:convert';
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:firebase_auth/firebase_auth.dart';
import 'package:firebase_core/firebase_core.dart';
import 'package:firebase_messaging/firebase_messaging.dart';
import 'package:flutter/foundation.dart';
import 'package:flutter_local_notifications/flutter_local_notifications.dart';
import 'package:invoify/core/helpers/notification_router.dart';

@pragma('vm:entry-point')
Future<void> firebaseMessagingBackgroundHandler(RemoteMessage message) async {
  await Firebase.initializeApp();
  debugPrint('Handling background message: ${message.messageId}');
}

class NotificationService {
  NotificationService({
    FirebaseMessaging? messaging,
    FirebaseFirestore? firestore,
    FirebaseAuth? auth,
    FlutterLocalNotificationsPlugin? localNotifications,
  }) : _messaging = messaging ?? FirebaseMessaging.instance,
       _firestore = firestore ?? FirebaseFirestore.instance,
       _auth = auth ?? FirebaseAuth.instance,
       _localNotifications =
           localNotifications ?? FlutterLocalNotificationsPlugin();

  final FirebaseMessaging _messaging;
  final FirebaseFirestore _firestore;
  final FirebaseAuth _auth;
  final FlutterLocalNotificationsPlugin _localNotifications;

  static const AndroidNotificationChannel _androidChannel =
      AndroidNotificationChannel(
        'high_importance_channel',
        'High Importance Notifications',
        description:
            'This channel is used for important invoice payment notifications.',
        importance: Importance.high,
      );

  Future<void> init() async {
    // 1. Setup Local Notification Channel for Android
    await _setupLocalNotifications();

    // 2. Request FCM Permission
    final settings = await _messaging.requestPermission(
      alert: true,
      badge: true,
      sound: true,
    );

    if (settings.authorizationStatus == AuthorizationStatus.authorized) {
      debugPrint('FCM Notification permission granted.');
      await _saveFcmToken();
    }

    // 3. Token refresh listener
    _messaging.onTokenRefresh.listen((token) async {
      await _saveTokenToFirestore(token);
    });

    // 4. Foreground message listener (Shows heads-up banner via Local Notifications)
    FirebaseMessaging.onMessage.listen((RemoteMessage message) {
      debugPrint(
        'Foreground FCM message received: ${message.notification?.title}',
      );
      _showForegroundNotification(message);
    });

    // 5. Background notification tap listener
    FirebaseMessaging.onMessageOpenedApp.listen((RemoteMessage message) {
      _handleNotificationTap(message);
    });

    // 6. Terminated state notification tap listener
    final initialMessage = await _messaging.getInitialMessage();
    if (initialMessage != null) {
      _handleNotificationTap(initialMessage);
    }
  }

  Future<void> _setupLocalNotifications() async {
    const androidInit = AndroidInitializationSettings('@mipmap/launcher_icon');
    const iosInit = DarwinInitializationSettings();
    const initSettings = InitializationSettings(
      android: androidInit,
      iOS: iosInit,
    );

    await _localNotifications.initialize(
      settings: initSettings,
      onDidReceiveNotificationResponse: (response) {
        final payload = response.payload;
        if (payload != null && payload.isNotEmpty) {
          NotificationRouter.handleInvoiceNavigation(payload);
        }
      },
    );

    await _localNotifications
        .resolvePlatformSpecificImplementation<
          AndroidFlutterLocalNotificationsPlugin
        >()
        ?.createNotificationChannel(_androidChannel);
  }

  void _showForegroundNotification(RemoteMessage message) {
    final notification = message.notification;
    final android = message.notification?.android;
    final invoiceId = message.data['invoiceId'] as String?;

    if (notification != null && !kIsWeb) {
      final payload = message.data.isNotEmpty
          ? jsonEncode(message.data)
          : (invoiceId ?? '');

      _localNotifications.show(
        id: notification.hashCode,
        title: notification.title,
        body: notification.body,
        notificationDetails: NotificationDetails(
          android: AndroidNotificationDetails(
            _androidChannel.id,
            _androidChannel.name,
            channelDescription: _androidChannel.description,
            importance: Importance.max,
            priority: Priority.high,
            icon: android?.smallIcon ?? '@mipmap/launcher_icon',
          ),
          iOS: const DarwinNotificationDetails(
            presentAlert: true,
            presentBadge: true,
            presentSound: true,
          ),
        ),
        payload: payload,
      );
    }
  }

  Future<void> _saveFcmToken() async {
    try {
      final token = await _messaging.getToken();
      if (token != null) {
        await _saveTokenToFirestore(token);
      }
    } catch (e) {
      debugPrint('Error getting FCM token: $e');
    }
  }

  Future<void> _saveTokenToFirestore(String token) async {
    final user = _auth.currentUser;
    if (user != null) {
      final language = PlatformDispatcher.instance.locale.languageCode;
      await _firestore.collection('users').doc(user.uid).set({
        'fcmToken': token,
        'languageCode': language,
        'lastTokenUpdate': FieldValue.serverTimestamp(),
      }, SetOptions(merge: true));
      debugPrint(
        'FCM Token & langCode ($language) saved to Firestore for user: ${user.uid}',
      );
    }
  }

  Future<void> updateLanguageCode(String languageCode) async {
    final user = _auth.currentUser;
    if (user != null) {
      await _firestore.collection('users').doc(user.uid).set({
        'languageCode': languageCode,
      }, SetOptions(merge: true));
      debugPrint(
        'Language code ($languageCode) updated in Firestore for user: ${user.uid}',
      );
    }
  }

  void _handleNotificationTap(RemoteMessage message) {
    if (message.data.isNotEmpty) {
      NotificationRouter.handleInvoiceNavigation(message.data);
    }
  }
}
